File: LeetCode/Medium/1472/1472.py
class Node:
    def __init__(self, data, forward=None, back=None):
        self.data = data
        self.forward = forward
        self.back = back


class BrowserHistory:
    root = None

    def __init__(self, homepage: str):
        BrowserHistory.root = Node(homepage)

    def visit(self, url: str) -> None:
        new_node = Node(url)
        new_node.back = BrowserHistory.root
        BrowserHistory.root.forward = new_node
        BrowserHistory.root = new_node

    def back(self, steps: int) -> str:
        cur_step = 0

        while BrowserHistory.root.back != None:
            cur_step += 1

            BrowserHistory.root = BrowserHistory.root.back

            if cur_step == steps:
                break

        return BrowserHistory.root.data

    def forward(self, steps: int) -> str:
        cur_step = 0

        while BrowserHistory.root.forward != None:
            cur_step += 1

            BrowserHistory.root = BrowserHistory.root.forward

            if cur_step == steps:
                break

        return BrowserHistory.root.data


# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)

File: LeetCode/Medium/1472/1472.java
class Node {

    String url;
    Node next;
    Node before;

    Node() {
        url = null;
        next = null;
        before = null;
    }
    
    Node(String url) {
        this.url = url;
    }
}
class BrowserHistory {

    Node homePage = null;

    public BrowserHistory(String homepage) {
        homePage = new Node(homepage);
    }
    
    public void visit(String url) {
        Node newPage = new Node(url);
        newPage.before = homePage;

        homePage.next = newPage;
        homePage = newPage;
    }
    
    public String back(int steps) {
        int move = 0;

        while(homePage.before != null) {
            homePage = homePage.before;
            move++;

            if (move == steps) {
                break;
            }
        }

        return homePage.url;
    }
    
    public String forward(int steps) {
        int move = 0;

        while(homePage.next != null) {
            homePage = homePage.next;
            move++;

            if (move == steps) {
                break;
            }
        }

        return homePage.url;
    }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.visit(url);
 * String param_2 = obj.back(steps);
 * String param_3 = obj.forward(steps);
 */